openmrs/omod/src/main/java/org/openmrs/projectbuendia/webservices/rest/EncounterResource.java

```diff
@@ -39,43 +40,34 @@
  * point in time (referred to as the "snapshot time").
  * @see AbstractReadOnlyResource
  */
-// TODO: Merge with PatientResource and let clients use a query parameter to
-// indicate whether encounter data should be included with each returned patient.
 @Resource(name = RestController.REST_VERSION_1_AND_NAMESPACE + "/encounters",
     supportedClass = Patient.class, supportedOpenmrsVersions = "1.10.*,1.11.*")
 public class EncounterResource
-    extends AbstractReadOnlyResource<Patient> implements Creatable {
+    extends AbstractReadOnlyResource<Encounter> implements Creatable {
     private final PatientService patientService;
-    private final EncounterService encounterService;
+    private final ProjectBuendiaService buendiaService;
 
     public EncounterResource() {
-        super("patient", Representation.DEFAULT);
+        super("encounter", Representation.DEFAULT);
         patientService = Context.getPatientService();
-        encounterService = Context.getEncounterService();
+        buendiaService = Context.getService(ProjectBuendiaService.class);
     }
 
     /**
-     * Returns all patients.  The retrieved records will be filled in with
-     * each patient's encounter and observation data by
-     * {@link #populateJsonProperties(Patient, RequestContext, SimpleObject, long)}
-     * on its way to becoming JSON that is sent to the client.
-     * @param context      unused here; see populateJsonProperties() for details
+     * Returns all encounters observed after the "since" parameter, or since the beginning of time
+     * if the "since" parameter is not set.
+     * @param context      used to obtain the "since" parameter for incremental sync.
      * @param snapshotTime unused here; see populateJsonProperties() for details
      * @see AbstractReadOnlyResource#search(RequestContext)
      */
-    @Override public List<Patient> searchImpl(RequestContext context, long snapshotTime) {
-        String patientUuid = context.getParameter("patient_uuid");
-        if (patientUuid != null) {
-            Patient patient = patientService.getPatientByUuid(patientUuid);
-            if (patient == null) {
-                throw new ObjectNotFoundException();
-            }
-            List<Patient> patients = new ArrayList<>();
-            patients.add(patient);
-            return patients;
-        } else {
-            return patientService.getAllPatients();
+    @Override public List<Encounter> searchImpl(RequestContext context, long snapshotTime) {
+        Date syncFrom;
+        try {
+            syncFrom = RequestUtil.getSyncFromDate(context);
```

Comment:
Oh, I understand what you're saying now! Sorry that it's taken a while.

Yes, the way I've described it above has a problem if:
- Tablet A is syncing data, and has 7 out of 10 pages
- Tablet B uploads a new encounter, with a UUID that falls within the first 70% of UUIDs on the system.

using `ORDER BY date, uuid` works around this, but will still result in the number of pages changing mid-sync. Which is ok using a `more` mechanism, and weird using a `num_pages` mechanism. Your idea of using [timestamp + uuid] as the bookmark doesn't suffer from this problem either.

There is also still an (extremely unlikely) edge case where this solution doesn't work.
- Records are being bulk added, so multiple records are being added with the same timestamp.
- Halfway through the bulk add, the client requests a page.
- More records are bulk added with the same timestamp
- The client requests another page, only to find that the pages are different.